src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -2443,104 +2441,38 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
-                    // then stream that to the stream that this method was called with. Otherwise output to the
-                    // stream that this method was called with.
-                    Stream retStream;
-                    if (!metadataOnly && IsRealSigned && this.Options.StrongNameProvider.Capability == SigningCapability.SignsStream)
-                    {
-                        Debug.Assert(Options.StrongNameProvider != null);
-
-                        try
-                        {
-                            signingInputStream = Options.StrongNameProvider.CreateInputStream();
-                        }
-                        catch (IOException e)
-                        {
-                            throw new Cci.PeWritingException(e);
-                        }
-
-                        retStream = signingInputStream;
-                    }
-                    else
-                    {
-                        signingInputStream = null;
-                        retStream = peStream;
-                    }
-
-                    return retStream;
+                    Stream ret;
+                    (peStream, signingInputStream, ret) = GetPeStream(metadataDiagnostics, peStreamProvider, metadataOnly);
+                    return ret;
                 };
 
-                Func<Stream> getRefPeStream;
-                if (metadataPEStreamProvider != null)
+                Func<Stream> getRefPeStream = () =>
                 {
-                    getRefPeStream = () =>
-                    {
-                        if (metadataDiagnostics.HasAnyErrors())
-                        {
-                            return null;
-                        }
+                    return ConditionalGetOrCreateStream(metadataPEStreamProvider, metadataDiagnostics);
+                };
 
-                        refPeStream = metadataPEStreamProvider.GetOrCreateStream(metadataDiagnostics);
-                        Debug.Assert(refPeStream != null || metadataDiagnostics.HasAnyErrors());
-                        return refPeStream;
-                    };
-                }
-                else
-                {
-                    getRefPeStream = null;
-                }
+                Func<Stream> getPortablePdbStream =
+                    moduleBeingBuilt.DebugInformationFormat != DebugInformationFormat.PortablePdb || pdbStreamProvider == null
+                    ? null
+                    : (Func<Stream>) (() => ConditionalGetOrCreateStream(pdbStreamProvider, metadataDiagnostics));
 
                 try
                 {
                     if (SerializePeToStream(
                         moduleBeingBuilt,
                         metadataDiagnostics,
-                        this.MessageProvider,
+                        MessageProvider,
                         getPeStream,
-                        getRefPeStream,
+                        metadataPEStreamProvider != null ? getRefPeStream : null,
```

Comment:
Consider moving `metadataPEStreamProvider != null` check to `getRefPeStream` declaration (similar to `getPortablePdbStream`) so we only allocate a delegate when needed.

Reply:
Done.